### ISSUE TITLE
QuickFix entries are incorrect

### ISSUE THREAD

**Issue by hubanya** [lervag/vimtex#421]
After compiling a `tex` file that contains errors, the quickfix window pops up. However, `:cn` and `:cp` does not correctly navigate through the entries in quickfix. In fact, it opens a new buffer.

Here are the steps to reproduce this problem on Windows.

Step 1: open gvim by double click gvim icon and open a `tex` file that contains two errors.
Step 2: compile using `\ll`, then quickfix window opens.
Step 3: use `:cn` to navigate to the next error. Instead of pointing to the error, gvim opens a new buffer.

Remarks:
1. The way one opens the `tex` file matters. If one uses something like `cmd`, `gvim c:\test.tex`, the above problem disappears.
2. Adding `let g:vimtex_quickfix_fix_paths=1` does not solve the problem.
3. Adding `set autochdir` solves the problem.
4. The above problem can also be reproduced on Ubuntu, again as long as the file is not opened from the command line `gvim ~/test.tex` for example.

**Comment by fuzzybear3965** [comment]
The issue seems to be related to the discrepancy between the working directory of a buffer (`echo getcwd()`) and the path of the file indicated in the log.

The LaTeX log file uses the path relative to the directory that the "shell" was in when latexmk was launched. `autochdir` makes the location of a file in the buffer the working directory for that buffer. vimtex, right now, changes the directory from where latexmk is launched to that which contains the file being edited. These paths are consistent as long as `autochdir` is set. However, if `autochdir` is not set then the path in which latexmk is started is different than the working directory of the LaTeX file's buffer. Since the log file doesn't contain the relative path from the working directory of the buffer to the directory that contains the LaTeX file, using the quickfix window to navigate the error list spawns a new buffer/file in the current working directory whose name is the same as that which was compiled in some other directory.

Note that this problem can be avoided by:

1) `set autochdir`
2) Explicitly setting the working directory to that of the file whose errors are listed in the quickfix window.
3) Changing vimtex so that it compiles from the working directory.

Note that none of these solutions have the advantage that if the working directory is changed after compilation that the quickfix window will necessarily work as expected. It may, therefore, be desirable to ensure that the working directory of the buffer at compile time and after compile time is the same.

**Comment by lervag** [comment]
First: Since the current version runs `latexmk` always in the same folder as the main LaTeX file, we know that the paths in the log file are relative to this path. Thus, if we prepend the main LaTeX file path to the quickfix filenames, the problem is solved. This is exactly what the `s:fix_quickfix_paths` do (and it is used as long as the option `g:vimtex_quickfix_fix_paths` is enabled. As far as I know, this works as expected on Linux. However, it doesn't work on Windows. The reason is, I believe, because of the explicit use of a slash in the pathname. Thus, I think this problem can be solved if I can update `s:fix_quickfix_paths` to work in all cases. In this case, perhaps it is best to make this mandatory and remove the option..?

**Comment by fuzzybear3965** [comment]
I think that's a good idea. I have been trying to make this work by compiling from the working directory. That way, the log has the right relative path. However, I'm having a lot of problems dealing with properly escaping these quotation marks. Also, if the user changes path after they have compiled then my approach becomes useless. It's better to have the quickfix window reference the absolute path of the document. I vote for removing the option.

**Comment by fuzzybear3965** [comment]
Okay, I think I have a patch for this. I'll submit my pull request shortly.

**Comment by lervag** [comment]
I think my latest commit should fix this. I found a couple of problems, in particular that I used the wrong autocmd event (should use `QuickFixCmdPost`, not `QuickFixCmdPre`).

I am still worried about the forward slash `/` that I use in `s:fix_quickfix_paths`, so I am not sure it fixes things on windows. At least, now it works as expected on Linux.

I'm curious if this should perhaps not be optional: It should not do any harm when it is not needed, but when people need this, it might be surprising that it is not enabled by default..?

**Comment by hubanya** [comment]
@lervag @fuzzybear3965  I just updated vimtex and tested it on Windows. I can confirm that it really works as expected now!  Thank you very much for your efforts!

**Comment by fuzzybear3965** [comment]
@hubanya, this is interesting because this update has not actually fixed the problem for me. I have made a minimal vimrc file that I would like you to use to test this. Please place this in `C:\Users\huban` (or your home directory, if that is not it); the file contents are below. I also have a `test.tex` file that I would like you to place in your `C:\Users\huban\Desktop` folder (or equivalent). Run the following commands, below.

1. In command line: `C:\Users\huban>gvim -u test.vimrc`
1. In gvim: `:e Desktop\test.tex`
1. In gvim: `:VimtexCompile`
3. In gvim: <kbd>Ctrl-w></kbd><kbd>j</kbd><kbd>return</kbd>

In my case, with this sequence of steps and the below configuration and test file in the above specified locations, I jump to the equivalent of `C:\Users\huban\test.tex`. @hubanya, can reply with the output of `:echo expand('%:p')` after executing the above sequence of commands? To make this easy for yourself, execute the following sequence of commands to place the text in your system clipboard (you can then <kbd>Ctrl-v</kbd> the result here):

1. `:redir @+ | echo expand('%:p') | redir end`

or, if you want to get really fancy, just copy the above to your system clipboard (omitting the leading colon), <kbd>Ctrl-c</kbd>, and then type the following in gvim, `:`<kbd>Ctrl-r</kbd>+<kbd>return/enter</kbd>. Note that the `+` is to be type literally. It refers to the system clipboard register.

----------------
#### min.vimrc

@lervag Maybe you can include this vimrc file in your issue/pull request template since it's prettty minimal.

```
set nocompatible
filetype plugin on
set rtp=$HOME/vimfiles/bundle/vimtex
set rtp+=$HOME/vimfiles/bundle/vimtex/after
let g:vimtex_view_general_viewer='SumatraPDF'
```
Visible, also, at: https://gist.github.com/fuzzybear3965/0b17adc602c0ace2f80ea6134a64bc90

---------------------------
#### ~\Desktop\test.tex
```
\documentclass{article}
\begin{document}
$ \alphab $
\end{document}
```

**Comment by lervag** [comment]
@fuzzybear3965 Just curious: Did you forget to set `let g:vimtex_quickfix_fix_paths = 1`? Based on the above mentioned minimal vimrc file, it seems you did...? Also, I [already do provide a minimal vimrc in the issue template](https://github.com/lervag/vimtex/blob/master/ISSUE_TEMPLATE.md). Mine doesn't override the `&rtp`, it only prepends and appends vimtex.

**Comment by lervag** [comment]
@hubanya I'm very happy to hear it works!

**Comment by fuzzybear3965** [comment]
Yeah, this works :/. I'm sorry that I forgot the whole point of the discussion. So, with regards to your point above, I see no reason not to make it the default. In the case where `autochdir` is on, issue #421 never happens. In the case where it's off (`set noautochdir`), the path is fixed properly. 

Also, your issue template is great. Don't touch it. Lesson learned: I need to be more thorough before I post suggestions.

**Comment by hubanya** [comment]
@fuzzybear3965 Sorry I'm late. But I'm glad to know that it also worked for you. Thank you very much for your efforts!

**Comment by lervag** [comment]
@fuzzybear3965 Thanks, great to know that it now works as expected! I'll remove the option entirely, as I want to keep the number of options low. If anyone ever finds a need for such an option, I'll consider to return it.

**Comment by lervag** [comment]
@fuzzybear3965 Btw: No problem, I'm happy for all the input and effort! :) And good to hear the there is no problem with the forward slashes. Btw, the use of `simplify` is really a substitute for my earlier `substitute` call. `simplify` will remove things like `./`, which was exactly what I manually did previously.

**Comment by fuzzybear3965** [comment]
Right, I had seen the replacement of the regex with `simplify` (the design is getting even cleaner!). But, I guess my post was half question and half monologue. If you'll notice, I said that the path in the quickfix window is only delimited with backslashes. This is strange because you [added a forward slash explicitly to the path](https://github.com/lervag/vimtex/commit/0f59eb086a701d8662129acd790c2cfa3eee57b1#diff-354d6427c13a97383c944d9dda674f40R548). My tests have confirmed that `simplify()` in conjunction with `fnamemodify(,':p')` should not "clean up" the path to be delimeted only by back slashes (`set noshellescape`). Do you clean up the path somewhere else?

**Comment by lervag** [comment]
No. I think this is something that is handled automatically by Vim. See `:h dos-backslash`. It seems Vim accepts forward slashes, and it seems it also accepts a mix of forward and backward slashes. So, my guess is that the path is never cleaned, but that when it is used by Vim, e.g. with `filereadable`, it automatically treats the slashes in the right way. I must admit that I don't really understand this that well. I guess one could read the Vim source code..?

In any case, the important thing is that it works! :)

**Comment by fuzzybear3965** [comment]
Hm, I'm not sure that `filereadable` does anything to change the path as seen in the quickfix window. I also didn't see anything in `:h dos-backslash` that clears up how vim might change one forward slash in a path to a backslash (only how it will take two backslashes and convert them to one). I really don't understand how vim escapes these backslashes. It seems that the use of backslashes is entirely avoidable so, in my opinion, vimtex should enable shellslash on a buffer which loads vimtex and/or vimtex should clean up backslash paths to completely avoid issues with escaping in the future (i.e. `vimtex#clean#paths(path)`. I think that that would make dealing with paths in the future a lot more predictable.

**Comment by lervag** [comment]
No, `filereadable` ovbiously doesn't affect the paths in the quickfix window. My point was that if `filereadable` can actually read files that mix forward and backward slashes (on windows), then perhaps this implies that Vim in general handles this mix for other commands. The command that changes the quickfix paths is `execute 'badd' l:file`, combined with changing the buffer number of the quickfix entry. Thus, it seems that `badd` can handle file names that mix forward and backward slashes. How this is done and why this is allowed beats me, but my main point above was that I don't really care, as long as it works :)

**Comment by lervag** [comment]
Regarding your last comment: I would agree, except it does not seem to be necessary. If I understood correctly, everything works as expected now. Or perhaps I am missing something?

If I am missing something, and if there are still issues with how paths are handled on Windows, then please, feel free to open a new issue and explain what the problem is.

**Comment by fuzzybear3965** [comment]
It is great when things work! It can be a problem, though, to adopt an implementation without full knowledge of how that implementation is functioning. If a change is made to the implementation later, it is possible that it will break if the inner workings of the black box aren't clear. For example, right now, issue #424 has arisen in response to #382. If it was fully understood how backslashes are escaped within vim and, then, with the windows command prompt issue #424 may never have arisen.

Simpler designs always result in designs that are easier to debug. I will issue a pull request for a `vimtex#clean#paths(path)` function that will convert backslashes to forward slashes. This should make paths independent of the shellslash property on Windows and make paths more consistent between operating systems.

**Comment by lervag** [comment]
I agree with keeping things simple and with using a clever design. But in this particular case, I am not quite sure if your proposal really solves any issues. As far as I can understand, the only issue right now is the issue with shellslash, and I think this can be solved by temporarily unsetting it before starting compilation and similar commands.

**Comment by lervag** [comment]
By the way, if you still want to try to improve how paths are handled: I think it would be better to put functions in `vimtex#util#clean_paths`. I also have `vimtex#util#shellescape` as an early attempt at trying to solve a similar issue.

Still, just to be clear: If we are able to find a better way of handling paths and shell command escaping, then it would really be very good. A major problem for me in this regard is that I don't use Vim and LaTeX on Windows, ever. I have a windows machine that I have tested on sometimes, but it is foreign to me and it takes a lot of time to work on for me. Thus, all of these problems are problems that I really never experience myself, and solving them depends on feedback and help from users.
